### PR TITLE
ENH: Add densify ufunc

### PR DESCRIPTION
**Pull request by brendan-ward**
This adds `GEOSDensify`, which will be available when GEOS 3.10.0 is released; currently available on GEOS `master`.

crossref to Shapely issue: https://github.com/Toblerity/Shapely/issues/1037

/cc @snowman2

### CODE REVIEW

File: CHANGELOG.rst [hunk 1]
```diff
@@ -11,14 +11,16 @@ Version 0.10 (unreleased)
...
-  under normal conditions.
+  under normal conditions (#286)
 
 **Added GEOS functions**
 
-* Addition of a ``contains_properly`` function (#267).
-* Addition of a ``polygonize`` function (#275).
+* Addition of a ``contains_properly`` function (#267)
+* Addition of a ``polygonize`` function (#275)
+* Addition of a ``densify`` function (#299)
```
Comment (jorisvandenbossche) [inline CHANGELOG.rst:20]: maybe mention that it's for GEOS 3.10+ only?

**Review by jorisvandenbossche — COMMENTED** [review]
PostGIS calls this `segmentize` (https://postgis.net/docs/ST_Segmentize.html, which I personally find a clearer name). Was there any discussion on the GEOS PR about naming?

**Comment by brendan-ward** [comment]
There was no discussion in the GEOS PR about naming.  It exposes operations against the `Densifier` C++ class (same naming in JTS as well), so I think "densify" was aligned with those terms, whereas "segmentize" would have been foreign.

Personally, I find "segmentize" less clear since "densify" implies increasing the density of points (in a linear geometry), whereas "segmentize" (i.e., make into segments) implies breaking a LineString into segments and returning individual segments (also would be useful later on).

**Comment by brendan-ward** [comment]
There was no discussion in the GEOS PR about naming.  It exposes operations against the `Densifier` C++ class (same naming in JTS as well), so I think "densify" was aligned with those terms, whereas "segmentize" would have been foreign.

Personally, I find "segmentize" less clear since "densify" implies increasing the density of points (in a linear geometry), whereas "segmentize" (i.e., make into segments) implies breaking a LineString into segments and returning individual segments (also would be useful later on).

**Comment by caspervdw** [comment]
Naming things is one of the hardest parts. In my opinion, both names are not concise, but that is the problem if you only have one word. Next best thing is aligning to what people know.

I'd like to follow PostGIS, this is what we have always done in pygeos (e.g. line_interpolate_point). End users do not care about namings of internal C++ classes, while they will probably know PostGIS.

**Review by caspervdw — COMMENTED** [review]
Looks good, only the name of the function is still under discussion

**Comment by brendan-ward** [comment]
Possible consistency issue: `frechet_distance` and `hausdorff_distance` include a `densify` parameter that looks to do something similar to this stand-alone function:
> The parameter 'densify' makes this approximation less coarse by splitting
    the line segments between vertices before computing the distance.

Does that make us want to rethink changing `densify` => `segmentize`?

**Comment by snowman2** [comment]
Side note: This closes #258. I created duplicate issues in pygeos and shapely.

**Review by caspervdw — APPROVED** [review]

**Pull request merged**